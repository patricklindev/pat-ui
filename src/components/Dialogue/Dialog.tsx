import React, { FC, MouseEvent } from 'react';
import { classNames } from '../../utils/classNames';

// export enum ButtonSize {
//   Large = 'lg',
//   Small = 'sm',
// }
export type ButtonSize = 'lg' | 'sm';
export type ButtonType =
  | 'primary'
  | 'secondary'
  | 'danger'
  | 'default'
  | 'link';
// export enum ButtonType {
//   Primary = 'primary',
//   Secondary = 'secondary',
//   Danger = 'danger',
//   Default = 'default',
//   Link = 'link',
// }
export interface IDialogProps {
  /** set customized style */
  className?: string;
}

type NativeButtonProps = IDialogProps;
type NativeAchorButtonProps = IDialogProps;
export type PatDialogProps = NativeButtonProps | NativeAchorButtonProps;

/**
 * A Button indicates a possible user action.
 *
 * ```js
 * import {Button} from 'pat-ui'
 * ```
 */
export const Dialog: FC<PatDialogProps> = (props) => {
  const { children, className, ...rest } = props;
  let styleClasses = 'dialog';
  if (className) styleClasses += ' ' + className;

  let dialog;
  dialog = <div className={styleClasses}>{props.children}</div>;
  return dialog;
};

Dialog.defaultProps = {};

export default Dialog;
