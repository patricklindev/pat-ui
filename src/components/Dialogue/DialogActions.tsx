import React, { FC } from 'react';
export interface IDialogActionsProps{
	/** set style classes*/
	className?:string;
	/** set whether to space child components */
	disableSpacing?:boolean;
}
export type PatDialogActionsProps = IDialogActionsProps

const DialogActions:FC<PatDialogActionsProps> = (props)=>{
const { disableSpacing, children, className } = props;

let dialogActions = <div className={`dialog__actions `+className} data-testid='dialog-actions-element'>{children}</div>
if(disableSpacing) {
	dialogActions.style.gap = 0;
}
return dialogActions;
}
DialogActions.defaultProps = {
	disableSpacing:false
}
export default DialogActions;