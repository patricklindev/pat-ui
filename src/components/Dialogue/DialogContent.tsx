import React, { FC } from 'react';
export interface IDialogContentProps{
	/** set style classes*/
	className?:string;
	/** set dividers option */
	dividers?:boolean;
}
export type PatDialogContentProps = IDialogContentProps
const DialogContent:FC<PatDialogContentProps> = (props)=>{
let classes = 'dialog__content';
const { dividers, children, className } = props;
if(className) classes += ' ' + className;
	return dividers ? 
	(<div className={classes} data-testid='dialog-content-element'><hr />{children}<hr /></div>) :
	(<div className={classes} data-testid='dialog-content-element'>{children}</div>)
}

DialogContent.defaultProps = {
	dividers:false
}
export default DialogContent;