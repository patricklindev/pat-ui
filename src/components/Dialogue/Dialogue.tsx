import React, { FC, useEffect, useRef } from 'react';
import { classNames } from '../../utils/classNames';

export interface IDialogueProps {
  /** set dialogue open */
  open: boolean;
  /** set onClose open */
  onClose: () => void;
  /** set customized dialogue */
  className?: string;
  /** set dialogue title */
  dialogueTitle?: string;
}

export type patDialogueProps = IDialogueProps;
/**
 * A Dialogs inform users about a task and can contain critical information, require decisions, or involve multiple tasks.
 *
 * A Dialog is a type of modal window that appears in front of app content to provide critical information or ask for a decision.
 * Dialogs disable all app functionality when they appear, and remain on screen until confirmed, dismissed, or a required action
 * has been taken.
 *
 * Dialogs are purposefully interruptive, so they should be used sparingly.
 * ```js
 * import {Dialogue} from 'pat-ui'
 * ```
 */
export const Dialogue: FC<patDialogueProps> = (props) => {
  const { open, onClose, className, dialogueTitle, children } = props;
  let styleClasses = classNames('dialogue');
  if (className) {
    styleClasses += ' ' + className;
  }
  const dialogRef = useRef(document.createElement('dialog'));

  // set open/close
  useEffect(() => {
    const dialogNode = dialogRef.current;
    if (open) {
      dialogNode.showModal();
    } else {
      dialogNode.close();
    }
  }, [open]);

  // Disable scroll when dialogue is open
  useEffect(() => {
    if (open) {
      document.body.style.overflow = 'hidden';
    } else {
      document.body.style.overflow = 'unset';
    }
  }, [open]);

  // Close dialogue when press ESC
  useEffect(() => {
    const close = (e: KeyboardEvent) => {
      if (e.keyCode === 27) {
        onClose();
      }
    };
    window.addEventListener('keydown', close);
    return () => window.removeEventListener('keydown', close);
  }, []);

  // Close dialogue when click outside
  useEffect(() => {
    dialogRef.current.addEventListener('click', (e) => {
      const node = e.target as HTMLElement;
      let rect = node.getBoundingClientRect();
      if (
        rect.left > e.clientX ||
        rect.right < e.clientX ||
        rect.top > e.clientY ||
        rect.bottom < e.clientY
      ) {
        onClose();
      }
    });
  }, []);

  let Dialogue = (
    <dialog className={styleClasses} ref={dialogRef}>
      <div className="dialogue-content">
        {dialogueTitle && (
          <div className="dialogue-title" data-testid="dialogue-title-element">
            <h4>{dialogueTitle}</h4>
          </div>
        )}
        <div className="dialogue-children">{children}</div>
      </div>
    </dialog>
  );

  return Dialogue;
};

export default Dialogue;
