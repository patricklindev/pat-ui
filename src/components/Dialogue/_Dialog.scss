.dialog {
  position: fixed; /* Stay in place */
  z-index: 1; /* Sit on top */
  left: 0;
  top: 0;
  overflow:hidden;
  width: 100%; /* Full width */
  height: 100%; /* Full height */
  background-color: rgb(0, 0, 0); /* Fallback color */
  background-color: rgba(0, 0, 0, 0.4); /* Black w/ opacity */
}
.dialog__content{
  margin: 15% auto;
  width:fit-content;
  overflow:hidden;
  background-color:white;
  padding:1em;
}
