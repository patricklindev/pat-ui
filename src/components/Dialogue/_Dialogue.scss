dialog::backdrop {
  background-color: rgba(0, 0, 0, 0.4);
}

.dialogue {
  position: fixed; /* Stay in place */
  z-index: 1; /* Sit on top */
  overflow: auto; /* Enable scroll if needed */
  box-sizing: content-box;
  background-color: #fefefe00;
  border: #fefefe00;
  max-width: 80%;
  padding: 0;

  .dialogue-content {
    border-radius: 5px;
    position: relative;
    background-color: #fefefe;
    margin: auto;
    padding: 15px;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
  }

  .dialogue-title {
    padding: 2px 16px;
  }
  .dialogue-children {
    padding: 2px 16px;
  }

  .dialogue-close {
    float: right;
    font-size: 28px;
    font-weight: bold;
    cursor: pointer;
  }
}
