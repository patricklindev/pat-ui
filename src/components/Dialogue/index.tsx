export { default } from './Dialog';
export { default } from './DialogTitle';
export { default } from './DialogContent';
import { default } from './DialogActions';