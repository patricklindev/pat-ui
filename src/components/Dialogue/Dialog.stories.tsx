import React, { useState} from 'react';
import { action } from '@storybook/addon-actions';
import Dialog from './Dialog';
import DialogTitle from './DialogTitle';
import DialogContent from './DialogContent';
import DialogContentText from './DialogContentText';
import DialogActions from './DialogActions';
import Icon from '../Icon';
import Button from '../Button';
import Input from '../Input';
import { DialogCloseReason } from './Dialog';
export default {
  title: 'Dialogue',
  component: Dialog,
};
interface SimpleDialogProps {
  open: boolean;
  selectedValue: string;
  onClose: (value: string) => void;
}
function SimpleDialogBody(props: SimpleDialogProps) {
  const { onClose, open, selectedValue } = props;
  const emails = ['username@gmail.com', 'user02@gmail.com'];
  const handleClose = () => {
    onClose(selectedValue);
  };

  const handleListItemClick = (value: string) => {
    onClose(value);
  };

  return (
    <Dialog onClose={handleClose} open={open} className="dialog">
      <DialogTitle>Set backup account</DialogTitle>
      <div>
      {emails.map((email) =>
          <div>
          <Icon name='plus' size='small' />
          <div>{email}</div>
          </div>)}
        <div>Add Account</div>  
      </div>
    </Dialog>
  );
}
export const SimpleDialog = ()=>{
  const emails = ['username@gmail.com', 'user02@gmail.com'];
  const [open, setOpen] = useState(false);
  const [selectedValue, setSelectedValue] = useState(emails[1]);
   const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = (value: string) => {
    setOpen(false);
    setSelectedValue(value);
  };
  return (
    <>
      <Button onClick={handleClickOpen}>
        Open simple dialog
      </Button>
      <SimpleDialogBody
        selectedValue={selectedValue}
        open={open}
        onClose={handleClose}
      />
    </>
    )
}

export const AlertDialog = () => {
  const [open, setOpen] = React.useState(false);

  const handleClickOpen = () => {
    action('opening dialog box!')();
    setOpen(true);
  };

  const handleClose = () => {
    action('closing dialog box!')();
    setOpen(false);
  };

  return (
    <div>
      <Button onClick={handleClickOpen}>
        Open alert dialog
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">
          {"Use Google's location service?"}
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Let Google help apps determine location. This means sending anonymous
            location data to Google, even when no apps are running.
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Disagree</Button>
          <Button onClick={handleClose} autoFocus>
            Agree
          </Button>
        </DialogActions> 
      </Dialog>
    </div>
  );
}

export const FormDialog = () => {
  const [open, setOpen] = useState(false);

  const handleClickOpen = () => {
    action('opening dialog box!')();
    setOpen(true);
  };

  const handleClose = () => {
    action('closing dialog box!')();
    setOpen(false);
  };

  return (
    <div>
      <Button onClick={handleClickOpen}>
        Open form dialog
      </Button>
      <Dialog open={open} onClose={handleClose} disableEscapeKeyDown={false}>
        <DialogTitle>Subscribe</DialogTitle>
        <DialogContent dividers={true}>
          To subscribe to this website, please enter your email address here. We
            will send updates occasionally.
          <Input />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Cancel</Button>
          <Button onClick={handleClose}>Subscribe</Button>
        </DialogActions>
      </Dialog>
    </div>
  );
}