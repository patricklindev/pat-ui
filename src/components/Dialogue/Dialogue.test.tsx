import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Dialogue from './Dialogue';
import Button from '../Button';
import Input from '../Input';
describe('Dialogue', () => {
  it('should match snapshot when not open', () => {
    // https://github.com/jsdom/jsdom/issues/3294
    HTMLDialogElement.prototype.close = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = false;
    });

    const { asFragment } = render(
      <Dialogue
        open={false}
        onClose={() => {}}
        className="jest-test-not-open"
        dialogueTitle="Jest Testing Not Open"
      >
        <p>Snapshot Dialogue</p>
      </Dialogue>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when open', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    const { asFragment } = render(
      <Dialogue
        open={true}
        onClose={() => {}}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>
          To subscribe to this website, please enter your email address here. We
          will send updates occasionally.
        </p>
        <Input fluid placeholder="Email Address" />
        <div
          style={{
            textAlign: 'right',
          }}
        >
          <Button
            btnType="link"
            disabled={false}
            onClick={function noRefCheck() {}}
          >
            Cancel
          </Button>
          <Button
            btnType="link"
            disabled={false}
            onClick={function noRefCheck() {}}
          >
            Subscribe
          </Button>
        </div>
      </Dialogue>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render dialogue', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    render(
      <Dialogue
        open={true}
        onClose={() => {}}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>Snapshot Dialogue</p>
      </Dialogue>
    );
    expect(screen.getByText('Snapshot Dialogue')).toBeInTheDocument();
  });

  it('should render dialogue with correct title', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    render(
      <Dialogue
        open={true}
        onClose={() => {}}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>Snapshot Dialogue</p>
      </Dialogue>
    );
    const dialogueTitleElement = screen.getByTestId('dialogue-title-element');
    expect(dialogueTitleElement).toHaveTextContent('Jest Testing Open');
  });

  it('should render dialogue with correct cancel button', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    render(
      <Dialogue
        open={true}
        onClose={() => {}}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>Snapshot Dialogue</p>
        <div
          style={{
            textAlign: 'right',
          }}
        >
          <Button
            btnType="link"
            disabled={false}
            onClick={function noRefCheck() {}}
            data-testid="dialogue-cancel-button-element"
          >
            Cancel
          </Button>
          <Button
            btnType="link"
            disabled={false}
            onClick={function noRefCheck() {}}
          >
            Subscribe
          </Button>
        </div>
      </Dialogue>
    );
    const dialogueCancelButtonElement = screen.getByTestId(
      'dialogue-cancel-button-element'
    );
    expect(dialogueCancelButtonElement).toBeInTheDocument();
  });

  it('should render dialogue with correct children', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    render(
      <Dialogue
        open={true}
        onClose={() => {}}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>Snapshot Dialogue</p>
      </Dialogue>
    );
    const dialogueChildrenElement = screen.getByText('Snapshot Dialogue');
    expect(dialogueChildrenElement).toBeInTheDocument();
  });

  it('should call onClose when close button is clicked', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    const onClose = jest.fn();
    render(
      <Dialogue
        open={true}
        onClose={onClose}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>Snapshot Dialogue</p>
        <Button
          btnType="link"
          disabled={false}
          onClick={onClose}
          data-testid="dialogue-close-button-element"
        >
          Close
        </Button>
      </Dialogue>
    );
    const dialogueCloseButtonElement = screen.getByTestId(
      'dialogue-close-button-element'
    );
    fireEvent.click(dialogueCloseButtonElement);
    expect(onClose).toHaveBeenCalled();
  });

  it('should call onClose when escape key is pressed', () => {
    HTMLDialogElement.prototype.showModal = jest.fn(function mock(
      this: HTMLDialogElement
    ) {
      this.open = true;
    });

    const onClose = jest.fn();
    render(
      <Dialogue
        open={true}
        onClose={onClose}
        className="jest-test-open"
        dialogueTitle="Jest Testing Open"
      >
        <p>Snapshot Dialogue</p>
      </Dialogue>
    );
    fireEvent.keyDown(document, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27,
    });
    expect(onClose).toHaveBeenCalled();
  });

});
