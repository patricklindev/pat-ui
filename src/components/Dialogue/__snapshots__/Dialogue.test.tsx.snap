// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogue should match snapshot when not open 1`] = `
<DocumentFragment>
  <dialog
    class="dialogue jest-test-not-open"
  >
    <div
      class="dialogue-content"
    >
      <div
        class="dialogue-title"
        data-testid="dialogue-title-element"
      >
        <h4>
          Jest Testing Not Open
        </h4>
      </div>
      <div
        class="dialogue-children"
      >
        <p>
          Snapshot Dialogue
        </p>
      </div>
    </div>
  </dialog>
</DocumentFragment>
`;

exports[`Dialogue should match snapshot when open 1`] = `
<DocumentFragment>
  <dialog
    class="dialogue jest-test-open"
    open=""
  >
    <div
      class="dialogue-content"
    >
      <div
        class="dialogue-title"
        data-testid="dialogue-title-element"
      >
        <h4>
          Jest Testing Open
        </h4>
      </div>
      <div
        class="dialogue-children"
      >
        <p>
          To subscribe to this website, please enter your email address here. We will send updates occasionally.
        </p>
        <div
          class="ui-input input-fluid"
        >
          <input
            placeholder="Email Address"
            type="text"
          />
        </div>
        <div
          style="text-align: right;"
        >
          <a
            class="btn btn-link"
          >
            Cancel
          </a>
          <a
            class="btn btn-link"
          >
            Subscribe
          </a>
        </div>
      </div>
    </div>
  </dialog>
</DocumentFragment>
`;
